stacks-signer/src/runloop.rs

```diff
@@ -0,0 +1,439 @@
+            }
+            (RunLoopCommand::DkgSign { message }, State::SignDone)
+            | (RunLoopCommand::Sign { message }, State::SignDone) => {
+                info!("Done Signing Message: {:?}", message);
+                self.state = State::Exit;
+            }
+            _ => {}
+        }
+    }
+
+    /// Process the event as both a signer and a coordinator
+    fn process_event(
+        &mut self,
+        event: &StackerDBChunksEvent,
+    ) -> (Vec<Message>, Vec<OperationResult>) {
+        //info!("Processing event: {:?}", event);
+        // Determine the current coordinator id and public key for verification
+        let (coordinator_id, coordinator_public_key) =
+            calculate_coordinator(&self.signing_round.public_keys);
+        // Filter out invalid messages
+        let inbound_messages: Vec<Message> = event
+            .modified_slots
+            .iter()
+            .filter_map(|chunk| {
+                let message = bincode::deserialize::<Message>(&chunk.data).ok()?;
+                if verify_msg(
+                    &message,
+                    &self.signing_round.public_keys,
+                    coordinator_public_key,
+                ) {
+                    Some(message)
+                } else {
+                    None
+                }
+            })
+            .collect();
+        // First process all messages as a signer
+        let mut outbound_messages =
+            process_inbound_messages(&mut self.signing_round, inbound_messages.clone())
+                .unwrap_or_default();
+        // If the signer is the coordinator, then next process the message as the coordinator
+        let (messages, results) = if self.signing_round.signer.signer_id == coordinator_id {
+            self.coordinator
+                .process_inbound_messages(inbound_messages)
+                .unwrap_or_default()
+        } else {
+            (vec![], vec![])
+        };
+        outbound_messages.extend(messages);
+        (outbound_messages, results)
+    }
+}
+
+impl RunLoop<FrostCoordinator> {
+    /// Creates new runloop from a config and command
+    pub fn new(config: &Config, command: RunLoopCommand) -> Self {
+        // TODO: should this be a config option?
+        // The threshold is 70% of the shares
+        let threshold = (config.signer_ids_public_keys.key_ids.len() / 10 * 7)
+            .try_into()
+            .unwrap();
+        let total_signers = config
+            .signer_ids_public_keys
+            .signers
+            .len()
+            .try_into()
+            .unwrap();
+        let total_keys = config
+            .signer_ids_public_keys
+            .key_ids
+            .len()
+            .try_into()
+            .unwrap();
+        let key_ids = config
+            .signer_key_ids
+            .get(&config.signer_id)
+            .unwrap()
+            .iter()
+            .map(|i| i - 1) // SigningRound::new (unlike SigningRound::from) doesn't do this
```

Comment:
This `- 1` can never underflow, right?

Reply:
It can absolutely underflow if we don't start our `key_ids` at `1`.  But we are starting at `1`, explicitly.

After I fix `wsts` to eliminate the whole off-by-one nastiness this will go away.